platforms/jvm/plugins-java/src/main/java/org/gradle/jvm/component/internal/DefaultJvmSoftwareComponent.java

```diff
@@ -186,4 +192,69 @@ public void withSourcesJar() {
     public JvmFeatureInternal getMainFeature() {
         return mainFeature;
     }
+
+    @Override
+    public void consistentResolution(Action<? super JavaResolutionConsistency> action, Project project) {
+        action.execute(new DefaultJavaResolutionConsistency(project, project.getExtensions().getByType(SourceSetContainer.class), project.getConfigurations()));
+    }
+
+    private static class DefaultJavaResolutionConsistency implements JavaResolutionConsistency {
+        private final Configuration mainCompileClasspath;
+        private final Configuration mainRuntimeClasspath;
+        private final Configuration testCompileClasspath;
+        private final Configuration testRuntimeClasspath;
+        private final SourceSetContainer sourceSets;
+        private final ConfigurationContainer configurations;
+
+        @Inject
+        public DefaultJavaResolutionConsistency(Project project, SourceSetContainer sourceSets, ConfigurationContainer configurations) {
+            this.sourceSets = sourceSets;
+            this.configurations = configurations;
+
+            if (project.getPlugins().hasPlugin(JavaPlugin.class)) {
```

Comment:
I think just outright copying this code into the component is not what we want here. 

We are _in_ the `DefaultJvmSoftwareComponent` here. The call to `JavaLibraryPlugin` below is unnecessary -- and one of the things we're trying to avoid. 

The `else` branch of this if statement should also live outside of the component. The else branch specifically handles the case where there is no jvm component to apply. It is deprecated legacy behavior that applies consistent resolution to "component-less" projects that are built manually with only source sets.

Reply:
I just pushed an update.  What do you think about this approach?

Reply:
I took a brief look. There are a lot more changes now than before. I wonder if it makes sense to have separate PRs for different things this PR is doing, since I think the scope of this PR has now become a lot bigger than the original `Migrate all consistent resolution logic to java component`

It's difficult to keep track of how this all interacts with each other, especially since some changes are not necessarily related to each other. I think it's simpler to reason about if we try to separate changes that are:
1. Purely moving files - These PRs are easy to review. They should be atomic and shouldn't update any code (except for maybe changing dependencies in a buildscript)
2. Refactoring code - These are more difficult to review than 1 but are relatively simple since we know refactoring should not change behavior. These should also be easier to review since comments should only be about style/structure and not behavior. 
3. Behavior changes - These require some discussion, decision making, deprecation warnings, and in general more thought on how things interact and how they affect users.

It seems we are doing all 3 of these things at once which is making this PR a bit hard to reason about. It seems like 4 separate things are happening here:
1. Moving consistent resolution to Java component
2. Updating how source/target compatibility is configured for jvm/groovy
3. Moving tests from :plugins to the JVM platform
4. Splitting out java-base plugin into its own subproject

I think we should do 1 and 2 (and maybe 3) separately before trying to continue with 4

---

_However_, to respond to your original comment, the consistent resolution code is doing a lot more than I originally thought it was. It is iterating over _all_ project source sets, even those that do not belong to the component, to make them resolve consistently with the component. I originally looked over this, and as such have probably misguided you on the solution. 

One feature should not be meddling with the source set of another feature. I was under the impression that the "consistency" being considered was local to a single component/feature. 

We might need to step back a bit and reconsider our approach. I'd have to think a bit more on how I'd rather this be, but to me if we want to keep the consistent resolution behavior the same, there would need to be some logic external to the component which links different components together. 

There could be a few solutions to this, including adding a _different_ interface to a component saying "here is a source set. Make this resolve consistently with you". Or, "here is a source set, resolve consistently with it". Then the logic to iterate over all sourcesets/components could live outside of the component. 

I think part of the problem here is that a source set is _kinda_ like a component, but really its only half of one. It defines the dependency scopes and resolvable configurations, but does not define the consumable parts of a component. Ive discussed this a bit in the past with Sterling, and we will definitely need to make some changes to components/source sets to help align these two similar concepts. 

It seems we have uncovered a stone that is a lot bigger than I originally imagined. I need to ponder a little more on this, but WDYT about meeting sometime early next week to discuss potential solutions?